### ISSUE TITLE
Why even use PyTorch's DataLoader in dataloader.py

### ISSUE THREAD

**Issue by LexSong** [nashory/pggan-pytorch#35]
It seems that the only way to get data from `self.dataloader`, is from `get_batch()`, and in `get_batch()` we recreate a new iterator of `self.dataloader` every time it is called. Therefore we always use only the first item returned by `self.dataloader`, which looks very pointless to create `self.dataloader` in the first place.

**Comment by akanimax** [comment]
Hi @LexSong, 

You are right about the `dataloader` always returning first element of the iterator. But since the dataloader while creation has set `shuffle=True`, It will return different data-points (well, hopefully :smile:). I wrote a package for pytorch for ProGANs (well structured code), you can check it out at [https://github.com/akanimax/pro_gan_pytorch](https://github.com/akanimax/pro_gan_pytorch). Examples for using this package are at [https://github.com/akanimax/pro_gan_pytorch-examples](https://github.com/akanimax/pro_gan_pytorch-examples) which validates that my code works.

Cheers!

**Comment by LexSong** [comment]
It's really inefficient though.

By default `RandomSampler` chooses samples without replacements, so it will generate a random indices list (by `torch.randperm(n).tolist()`) every time we call it. If randomly sampling with replacements is what we want here, just create a dataset properly.

**Comment by akanimax** [comment]
@LexSong, Yes, I know it is inefficient. BTW, I am not the author of this code, was just helping. Did you check out the links?

**Comment by shamanez** [comment]
@akanimax  your codes are perfect. I even followed the VDB code which is very efficient and informative.
